store/tikv/gcworker/gc_worker.go

```diff
@@ -1341,3 +1712,204 @@ func (w *MockGCWorker) DeleteRanges(ctx context.Context, safePoint uint64) error
+	lock := r.PeekNextLock()
+	r.NextLock = nil
+	return lock
+}
+
+// mergeReceiver is a list of receivers
+type mergeReceiver []*receiver
+
+func (r mergeReceiver) Len() int {
+	return len(r)
+}
+
+func (r mergeReceiver) Less(i, j int) bool {
+	lhs := r[i].PeekNextLock()
+	rhs := r[j].PeekNextLock()
+	// nil which means the receiver has finished should be the greatest one.
+	if lhs == nil {
+		// lhs >= rhs
+		return false
+	}
+	if rhs == nil {
+		// lhs != nil, so lhs < rhs
+		return true
+	}
+
+	return bytes.Compare(lhs.Key, rhs.Key) < 0
+}
+
+func (r mergeReceiver) Swap(i, j int) {
+	r[i], r[j] = r[j], r[i]
+}
+
+func (r *mergeReceiver) Push(x interface{}) {
+	*r = append(*r, x.(*receiver))
+}
+
+func (r *mergeReceiver) Pop() interface{} {
+	receivers := *r
+	res := receivers[len(receivers)-1]
+	*r = receivers[:len(receivers)-1]
+	return res
+}
+
+type scanLockResult struct {
+	Lock *tikv.Lock
+	Err  error
+}
+
+func newMergeLockScanner(safePoint uint64, client tikv.Client, stores map[uint64]*metapb.Store) *mergeLockScanner {
+	return &mergeLockScanner{
+		safePoint: safePoint,
+		client:    client,
+		stores:    stores,
+	}
+}
+
+// Start initializes the scanner and enables retrieving items from the scanner.
+func (s *mergeLockScanner) Start(ctx context.Context) error {
+	receivers := make([]*receiver, 0, len(s.stores))
+
+	for storeID, store := range s.stores {
+		ch := make(chan scanLockResult, scanLockResultBufferSize)
+		go func() {
+			err := s.physicalScanLocksForStore(ctx, s.safePoint, store, ch)
+			if err != nil {
+				logutil.Logger(ctx).Error("physical scan lock for store encountered error",
+					zap.Uint64("safePoint", s.safePoint),
+					zap.Any("store", store),
+					zap.Error(err))
+				ch <- scanLockResult{Err: err}
+			}
+		}()
+		receivers = append(receivers, &receiver{Ch: ch, StoreID: storeID})
+	}
+
+	s.startWithReceivers(receivers)
+
+	return nil
+}
+
+func (s *mergeLockScanner) startWithReceivers(receivers []*receiver) {
+	s.receivers = receivers
+	heap.Init(&s.receivers)
+}
+
+func (s *mergeLockScanner) Next() *tikv.Lock {
+	for {
+		nextReceiver := heap.Pop(&s.receivers).(*receiver)
+		nextLock := nextReceiver.TakeNextLock()
+		heap.Push(&s.receivers, nextReceiver)
+
+		if nextLock == nil {
+			return nil
+		}
+		if s.currentLockKey == nil || !bytes.Equal(s.currentLockKey, nextLock.Key) {
+			s.currentLockKey = nextLock.Key
+			return nextLock
+		}
+	}
+}
+
+func (s *mergeLockScanner) NextBatch(batchSize int) []*tikv.Lock {
+	result := make([]*tikv.Lock, 0, batchSize)
+	for len(result) < batchSize {
+		lock := s.Next()
+		if lock == nil {
+			return result
+		}
+		result = append(result, lock)
+	}
+	return result
+}
+
+// GetSucceededStores gets a set of successfully scanned stores. Only call this after finishing scanning all locks.
+func (s *mergeLockScanner) GetSucceededStores() map[uint64]interface{} {
+	stores := make(map[uint64]interface{})
+	for _, receiver := range s.receivers {
+		if receiver.Err == nil {
```

Comment:
If receiver has error, I think we should treat it just as an error?

Reply:
Did you mean returning an error? Here our solution is, for unsuccessful stores, GetSucceededStores doesn't return it and it will be scanned in next retry. The error is just logged.